source/endpoints_ruleset.c

```diff
@@ -0,0 +1,939 @@
+    }
+
+    struct aws_endpoints_parameter *parameter = aws_endpoints_parameter_new(wrapper->allocator, type, key);
+    parameter->documentation = aws_string_new_from_cursor(wrapper->allocator, &documentation_cur);
+
+    /* optional fields */
+    parameter->built_in = NULL;
+    struct aws_byte_cursor built_in_cur;
+    struct aws_json_value *built_in_node = aws_json_value_get_from_object(value, aws_byte_cursor_from_c_str("builtIn"));
+    if (built_in_node != NULL && !aws_json_value_get_string(built_in_node, &built_in_cur)) {
+        parameter->built_in = aws_string_new_from_cursor(wrapper->allocator, &built_in_cur);
+    }
+
+    parameter->is_required = false;
+    struct aws_json_value *required_node =
+        aws_json_value_get_from_object(value, aws_byte_cursor_from_c_str("required"));
+    if (required_node != NULL) {
+        aws_json_value_get_boolean(required_node, &parameter->is_required);
+    }
+
+    struct aws_json_value *default_node = aws_json_value_get_from_object(value, aws_byte_cursor_from_c_str("default"));
+    if (default_node != NULL) {
+        if (type == AWS_ENDPOINTS_PARAMETER_STRING) {
+            struct aws_byte_cursor default_cur;
+            if (aws_json_value_get_string(default_node, &default_cur)) {
+                aws_endpoints_parameter_destroy(parameter);
+                AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Unexpected type for default parameter value.");
+                return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+            }
+            parameter->default_value.string = aws_string_new_from_cursor(wrapper->allocator, &default_cur);
+        } else if (type == AWS_ENDPOINTS_PARAMETER_BOOLEAN) {
+            if (aws_json_value_get_boolean(default_node, &parameter->default_value.boolean)) {
+                aws_endpoints_parameter_destroy(parameter);
+                AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Unexpected type for default parameter value.");
+                return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+            }
+        }
+    }
+
+    struct aws_json_value *deprecated_node =
+        aws_json_value_get_from_object(value, aws_byte_cursor_from_c_str("deprecated"));
+    if (deprecated_node != NULL) {
+
+        struct aws_byte_cursor deprecated_message_cur;
+        struct aws_json_value *deprecated_message_node =
+            aws_json_value_get_from_object(deprecated_node, aws_byte_cursor_from_c_str("message"));
+        if (deprecated_message_node != NULL &&
+            aws_json_value_get_string(deprecated_message_node, &deprecated_message_cur)) {
+            aws_endpoints_parameter_destroy(parameter);
+            AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Unexpected value for deprecated message.");
+            return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+        }
+        parameter->deprecated_message = aws_string_new_from_cursor(wrapper->allocator, &deprecated_message_cur);
```

Comment:
bug:
- if there is no "message" node, then `deprecated_message_cur` is uninitialized
- therefore `parameter->deprecated_message = aws_string_new_from_cursor(wrapper->allocator, &deprecated_message_cur);` could end up as ... anything

probably want to break the if-statement into 2:
```
if deprecated_message_node != NULL:
    if aws_json_value_get_string(deprecated_message_node, &deprecated_message_cur):
        ERROR
    parameter->deprecated_message = aws_string_new_from_cursor(...)